carma-messenger-core/cpp_message/src/MobilityRequest_Message.cpp

```diff
@@ -0,0 +1,538 @@
+/*
+ * Copyright (C) 2020 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/**
+ * CPP File containing Mobility Request Message method implementations
+ */
+
+#include "MobilityRequest_Message.h"
+#include "MobilityHeader_Message.h"
+
+namespace cpp_message
+{
+    boost::optional<cav_msgs::MobilityRequest> Mobility_Request::decode_mobility_request_message(std::vector<uint8_t>& binary_array)
+    {
+        cav_msgs::MobilityHeader header;
+        cav_msgs::MobilityRequest output;
+
+        //decode results - stored in binary array
+        asn_dec_rval_t rval;
+        MessageFrame_t* message=nullptr;
+
+        //copy from vector to array
+        size_t len=binary_array.size();
+
+        uint8_t buf[len];
+        for(size_t i=0;i<len;i++){
+            buf[i]=binary_array[i];
+        }
+        
+        //use asn1c lib to decode
+        rval=uper_decode(0, &asn_DEF_MessageFrame, (void **) &message,buf, len,0,0);
+        if(rval.code==RC_OK){
+            Mobility_Header Header_constant;
+            std::string sender_id, recipient_id, sender_bsm_id, plan_id;
+            uint64_t timestamp;
+            //get sender id
+            size_t str_len=message->value.choice.TestMessage00.header.hostStaticId.size;
```

Comment:
It might be worth considering if it'd save any effort/be any cleaner to make a generic "MobilityHeader processor" or something to do this unpacking and checking in each message type. Maybe not since this is the last mobility message, but it's worth thinking on for a minute.

Reply:
I did attempt to that, but I had to return a pointer to an allocated memory for the encoded/decoded mobility header message and it kept getting reinitialized on calling return.